src/Compilers/CSharp/Portable/Symbols/SubstitutedTypeParameterSymbol.cs

```diff
@@ -99,8 +100,53 @@ public override ImmutableArray<CSharpAttributeData> GetAttributes()
         internal override ImmutableArray<TypeWithAnnotations> GetConstraintTypes(ConsList<TypeParameterSymbol> inProgress)
         {
             var constraintTypes = ArrayBuilder<TypeWithAnnotations>.GetInstance();
-            _map.SubstituteTypesDistinctWithoutModifiers(_underlyingTypeParameter.GetConstraintTypes(inProgress), constraintTypes, null);
-            return constraintTypes.ToImmutableAndFree().WhereAsArray(type => type.SpecialType != SpecialType.System_Object || !type.NullableAnnotation.IsAnnotated());
+            _map.SubstituteConstraintTypesDistinctWithoutModifiers(_underlyingTypeParameter.GetConstraintTypes(inProgress), constraintTypes, null);
+
+            TypeWithAnnotations bestObjectConstraint = default;
+
+            // Strip all Object constraints.
+            for (int i = constraintTypes.Count - 1; i >= 0; i--)
+            {
+                TypeWithAnnotations type = constraintTypes[i];
+                if (ConstraintsHelper.IsObjectConstraint(type, ref bestObjectConstraint))
+                {
+                    constraintTypes.RemoveAt(i);
+                }
+            }
+
+            if (bestObjectConstraint.HasType)
+            {
+                // See if we need to put Object! or Object~ back in order to preserve nullability information for the type parameter.
+                if (ConstraintsHelper.IsObjectConstraintSignificant(CalculateIsNotNullableFromNonTypeConstraints(), bestObjectConstraint))
+                {
+                    Debug.Assert(!HasNotNullConstraint && !HasValueTypeConstraint);
+                    if (constraintTypes.Count == 0)
+                    {
+                        if (bestObjectConstraint.NullableAnnotation.IsOblivious() && !HasReferenceTypeConstraint)
+                        {
+                            bestObjectConstraint = default;
+                        }
+                    }
+                    else
+                    {
+                        foreach (TypeWithAnnotations constraintType in constraintTypes)
+                        {
+                            if (!ConstraintsHelper.IsObjectConstraintSignificant(IsNotNullableIfReferenceTypeFromConstraintType(constraintType, out _), bestObjectConstraint))
+                            {
+                                bestObjectConstraint = default;
+                                break;
+                            }
+                        }
+                    }
+
+                    if (bestObjectConstraint.HasType)
+                    {
+                        constraintTypes.Insert(0, bestObjectConstraint);
```

Comment:
>.Insert(0,  [](start = 39, length = 11)

nit: why insert at index 0 rather than just append with `.Add`?

Reply:
> why insert at index 0 rather than just append with .Add?

Having significant constraint at the beginning is likely to make following calculations faster - we might get the answer from the first item and stop.

---
In reply to: [292127065](https://github.com/dotnet/roslyn/pull/36269#discussion_r292127065) [](ancestors = 292127065)